src/main/java/org/entando/kubernetes/service/digitalexchange/BundleUtilities.java

```diff
@@ -486,22 +487,72 @@ public static String determineSignedBundleFolder(BundleType bundleType, String d
      */
     public static List<EnvVar> assemblePluginEnvVars(List<EnvironmentVariable> environmentVariableList,
             List<EnvVar> customEnvironmentVariablesList) {
+        String message;
 
         Map<String, EnvVar> assembledEnvVar = new HashMap<>();
         Optional.ofNullable(environmentVariableList).ifPresent(l -> l.stream()
-                .map(env -> buildFromEnvironmentVariable(env)).forEach(env -> {
-                    assembledEnvVar.put(env.getName(), env);
-                }));
+                .map(BundleUtilities::buildFromEnvironmentVariable).forEach(env ->
+                    assembledEnvVar.put(env.getName(), env)
+                ));
+        // adding tenant code
+        final String currentTenantCode = TenantContextHolder.getCurrentTenantCode();
+
         assembledEnvVar.put(ENTANDO_TENANT_CODE, buildFromEnvironmentVariable(
-                new EnvironmentVariable(ENTANDO_TENANT_CODE, TenantContextHolder.getCurrentTenantCode(), null)));
-        Optional.ofNullable(customEnvironmentVariablesList).ifPresent(l -> l.stream().forEach(env -> {
-            assembledEnvVar.put(env.getName(), env);
-        }));
+                new EnvironmentVariable(ENTANDO_TENANT_CODE, currentTenantCode, null)));
 
-        return assembledEnvVar.entrySet().stream()
-                .map(Entry::getValue)
+        Optional.ofNullable(customEnvironmentVariablesList).ifPresent(l -> l.stream()
+                .forEach(env ->
+                    assembledEnvVar.put(env.getName(), env)
+                ));
+
+        List<EnvVar> list = assembledEnvVar.values().stream()
                 .sorted(Comparator.comparing(EnvVar::getName))
                 .collect(Collectors.toList());
+        // check for primary secrets
+        if (currentTenantCode.equals(PRIMARY_TENANT))  {
+            message = "Cannot reference a non-primary secret on the primary tenant!";
+        } else {
+            message = "Cannot reference a primary secret on the non-primary tenant '" + currentTenantCode + "'!";
+        }
+        // make sure the correct kind of secret name is referenced
+        list.stream().filter(e -> e.getValueFrom() != null
+                        && isErrorConditionMet(countHashesInSecretName(e.getValueFrom().getSecretKeyRef().getName(), ENTITY_CODE_HASH_LENGTH)))
+                .findFirst()
+                .ifPresent(v -> {
+                    throw new EntandoValidationException(message + "Faulting environment variable name '" + v.getName() + "'");
+                });
+        return list;
+    }
+
+    /**
+     * Make sure that the primary tenant got secrets with at least two hashes, three or more for non-primary tenants.
+     * @param count the number ho hashed detected in secret name
+     * @return true if there is an error, false otherwise
+     */
+    private static boolean isErrorConditionMet(Integer count) {
```

Comment:
Can we improve readability by using a more meaningful name here? Something like 

```
doesSecretBelongToCurrentTenant(String name) {
  int count = countHashesInSecretName(name, ENTITY_CODE_HASH_LENGTH);
  return (TenantContextHolder.getCurrentTenantCode().equals(PRIMARY_TENANT) && count > 2)
                || (!TenantContextHolder.getCurrentTenantCode().equals(PRIMARY_TENANT) && count < 3);
}
```
I've probably inverted the logic in my example, it's not needed obsviously, it's only an example

Please consider that `BundleUtilities` is a very generic purpose class, using a name like `isErrorConditionMet` could be misleading